pag-lexer/src/normalization.rs

```diff
@@ -6,182 +6,193 @@
+            }
+            let new: RcVec = new
+                .into_iter()
+                .filter(|x| !matches!(x.as_ref(), Epsilon))
+                .collect();
+            if new.len() == 1 {
+                new[0].clone()
+            } else if new.is_empty() {
+                RegexTree::epsilon()
+            } else if sequence_unchanged(&new, old) {
+                node
+            } else {
+                Rc::new(Concat(new))
             }
-            Rc::new(Concat(nr, ns))
         }
-        KleeneClosure(r) => {
-            let nr = normalize(r.clone());
-            match &*nr {
-                KleeneClosure(_) => nr,
-                Bottom => Rc::new(Epsilon),
-                Epsilon => Rc::new(Epsilon),
+        KleeneClosure(old) => {
+            let new = normalize(old.clone());
+            match &*new {
+                KleeneClosure(_) => new,
+                Bottom | Epsilon => RegexTree::epsilon(),
                 _ => {
-                    if Rc::ptr_eq(r, &nr) {
-                        return tree;
+                    if Rc::ptr_eq(old, &new) {
+                        node
+                    } else {
+                        Rc::new(KleeneClosure(new))
                     }
-                    Rc::new(KleeneClosure(nr))
                 }
             }
         }
-        Union(r, s) => {
-            // referential equality
-            if Rc::ptr_eq(r, s) {
-                return normalize(r.clone());
-            }
-            let r = normalize(r.clone());
-            let s = normalize(s.clone());
+        Union(old) => {
+            let new: RcVec = old.iter().map(|x| normalize(x.clone())).collect();
+            let new: RcVec = flatten_union(&new);
 
-            // empty | r = r
-            if matches!(&*r, Bottom) {
-                return s;
-            }
-            if matches!(&*s, Bottom) {
-                return r;
+            let mut sets = None;
+            let mut nonsets = Vec::new();
+
+            for i in new {
+                if let Set(x) = i.as_ref() {
+                    sets = match sets {
+                        None => Some(x.clone()),
+                        Some(y) => Some(x.union(&y)),
+                    }
+                } else {
+                    nonsets.push(i);
+                }
             }
-            // any | r = any
-            if matches!(&*s, Top) {
-                return s;
+
+            let mut new: RcVec = sets
+                .map(|x| Rc::new(Set(x)))
+                .into_iter()
+                .chain(nonsets.into_iter())
+                .collect();
+
+            new.sort_unstable();
+            new.dedup_by(|x, y| Rc::ptr_eq(x, y) || x == y);
+            new = new
+                .into_iter()
+                .filter(|x| !matches!(x.as_ref(), Bottom))
+                .collect();
+
+            if new
+                .iter()
+                .any(|x| !matches!(x.as_ref(), Epsilon) && x.is_nullable())
```

Comment:
What's the purpose of `x.is_nullable()` here?

Reply:
handle the cases of `eps | x*` => `x*`

Reply:
Fair